File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/presentation/ui/base/activity/BaseDrawerActivity.java
package com.lattis.ellipse.presentation.ui.base.activity;

import android.content.res.Configuration;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.navigation.NavigationView;

import java.util.List;

import butterknife.BindView;
import io.lattis.ellipse.R;

public abstract class BaseDrawerActivity<Presenter extends ActivityPresenter>
        extends BaseAuthenticatedActivity<Presenter> {

    @Nullable
    @BindView(R.id.navigation_view)
    NavigationView navigationView;
    @Nullable
    @BindView(R.id.navigation_recycler_view)
    RecyclerView recyclerView;
    @Nullable
    @BindView(R.id.drawer_layout)
    DrawerLayout drawerLayout;
    ActionBarDrawerToggle drawerToggle;

    private DrawerAdapter drawerAdapter;

    private List<DrawerMenu> drawerMenus;
    private Boolean isMenuShownForRide=false;


    @Override
    protected int getViewStubId() {
        return R.id.view_stub;
    }

    @Override
    protected int getViewStubLayoutId() {
        return getActivityContentLayoutId();
    }

    @Override
    protected void configureViews() {
        if(drawerLayout!=null && toolbar != null){
            drawerToggle = new ActionBarDrawerToggle(this, drawerLayout, toolbar, R.string.menu_open, R.string.menu_close);
            drawerLayout.addDrawerListener(drawerToggle);
            setupAppbar(toolbar,false);
            drawerToggle.syncState();
        }

        if(isMenuShownForRide)
            drawerMenus = DrawerMenu.getMenusWithRide();
        else
            drawerMenus = DrawerMenu.getMenusWithoutRide();


        drawerAdapter = new DrawerAdapter(drawerMenus);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(drawerAdapter);
    }





    @Override
    protected void setToolbarBackArrowAction() { }

    protected void setCheckedItem(DrawerMenu menuItem) {
        int newPosition=-1;
        for(int position=0;position<drawerMenus.size();position++){
            if(drawerMenus.get(position).getItemId() == menuItem.getItemId()){
                newPosition=position;
                break;
            }
        }
        if(newPosition==-1)
            return;

        //int newPosition = DrawerMenu.valueOf(menuItem.toString()).ordinal();

        drawerAdapter.notifyItemChanged(drawerAdapter.getSelectedPosition());
        drawerAdapter.setSelectedPosition(newPosition);
        drawerAdapter.notifyItemChanged(newPosition);
    }

    public void setHomeForBackPressed(){
        setCheckedItem(drawerMenus.get(0));
    }

    protected void navigateTo(DrawerMenu menuItem) {
        setCheckedItem(menuItem);
        onDrawerItemClicked(menuItem);
    }

    @Override
    protected void configureSubscriptions() {
        super.configureSubscriptions();
        if (navigationView != null) {
            subscriptions.add(drawerAdapter
                    .getViewClickSubject()
                    .subscribe(view -> {
                        int newPosition = recyclerView.getChildAdapterPosition(view);

                        drawerAdapter.notifyItemChanged(drawerAdapter.getSelectedPosition());
                        drawerAdapter.setSelectedPosition(newPosition);
                        drawerAdapter.notifyItemChanged(newPosition);

                        DrawerMenu itemMenu = drawerMenus.get(newPosition);
                        onDrawerItemClicked(itemMenu);
                    }));
        }
    }

    protected void onDrawerItemClicked(DrawerMenu menuItem){
        onDrawerItemSelected(menuItem);
        closeDrawer();
    }

    protected void onDrawerItemSelected(DrawerMenu menuItem){}

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        if(drawerToggle!=null){
            drawerToggle.syncState();
        }
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if(drawerToggle!=null){
            drawerToggle.onConfigurationChanged(newConfig);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (drawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        if(isDrawerOpen()){
            closeDrawer();
        } else {
            super.onBackPressed();
        }
    }

    private boolean isDrawerOpen(){
        return drawerLayout != null && navigationView!=null &&
                drawerLayout.isDrawerOpen(navigationView);
    }

    public void closeDrawer(){
        if(drawerLayout != null && navigationView!=null){
            drawerLayout.closeDrawer(navigationView);
        }
    }

    @Override
    protected int getActivityLayoutId() {
        return R.layout.activity_base_drawer;
    }

    protected abstract int getDefaultSelectedItemId();


    protected void setBaseDrawerMenuWithRide(){
        isMenuShownForRide=true;
        drawerMenus = DrawerMenu.getMenusWithRide();
        if(recyclerView!=null && drawerAdapter!=null){
            drawerAdapter.setDrawerMenuList(drawerMenus);
            drawerAdapter.notifyDataSetChanged();
        }

    }

    protected void setBaseDrawerMenuWithoutRide(){
        isMenuShownForRide=false;
        drawerMenus = DrawerMenu.getMenusWithoutRide();
        if(recyclerView!=null && drawerAdapter!=null){
            drawerAdapter.setDrawerMenuList(drawerMenus);
            drawerAdapter.notifyDataSetChanged();
        }
    }

    protected abstract int getActivityContentLayoutId();



    protected void onInternetConnectionChanged(boolean isConnected) {
        if(isConnected){
            drawerToggle.setDrawerIndicatorEnabled(true);
        }else{
            drawerToggle.setDrawerIndicatorEnabled(false);
        }
        drawerToggle.syncState();

    }
}

File: velolabs_repos/lattis_android_v2/app/src/main/java/com/lattis/lattis/presentation/base/activity/BaseDrawerActivity.kt
package com.lattis.lattis.presentation.base.activity

import android.content.res.Configuration
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.LinearLayoutManager
import com.lattis.domain.models.Reservation
import com.lattis.lattis.presentation.base.BaseView
import com.lattis.lattis.presentation.base.activity.DrawerAdapter.Companion.ITEM
import io.lattis.lattis.R
import kotlinx.android.synthetic.main.view_drawer_content.*
import kotlinx.android.synthetic.main.view_drawer_layout.*
import kotlinx.android.synthetic.main.view_toolbar.*

abstract class BaseDrawerActivity<Presenter : ActivityPresenter<V>,V:BaseView> : BaseAuthenticatedActivity<Presenter,V>() {

    lateinit var drawerToggle: ActionBarDrawerToggle
    private lateinit var drawerAdapter: DrawerAdapter
    private lateinit var drawerMenus: List<DrawerMenu>
    private var isMenuShownForRide = false
    override val viewStubIdFullScreen
        protected get() = R.id.view_stub_full_screen

    protected abstract val activityContentLayoutId: Int
    override val viewStubLayoutId: Int = activityContentLayoutId

    override fun configureViews() {
        if (drawer_layout != null && toolbar != null) {
            drawerToggle = ActionBarDrawerToggle(
                this,
                drawer_layout,
                toolbar,
                R.string.menu_open,
                R.string.menu_close
            )
            drawer_layout?.addDrawerListener(drawerToggle)
            setupAppbar(toolbar, false)
            drawerToggle?.syncState()
        }
        drawerMenus =
            if (isMenuShownForRide) DrawerMenu.Companion.menusWithRide() else DrawerMenu.Companion.menusWithoutRide()
        drawerAdapter = DrawerAdapter(drawerMenus)
        navigation_recycler_view?.layoutManager = LinearLayoutManager(this)
        navigation_recycler_view?.adapter = drawerAdapter
    }

    override fun setToolbarBackArrowAction() {}


    protected fun setCheckedItem(menuItem: DrawerMenu) {
        var newPosition = -1
        for (position in drawerMenus?.indices) {
            if (drawerMenus[position].itemId == menuItem.itemId) {
                newPosition = position
                break
            }
        }
        if (newPosition == -1) return
        //int newPosition = DrawerMenu.valueOf(menuItem.toString()).ordinal();
        drawerAdapter?.notifyItemChanged(drawerAdapter.selectedPosition)
        drawerAdapter?.selectedPosition = newPosition
        drawerAdapter?.notifyItemChanged(newPosition)
    }

    fun setHomeForBackPressed() {
        setCheckedItem(drawerMenus[0])
    }

    protected fun navigateTo(menuItem: DrawerMenu) {
        setCheckedItem(menuItem)
        onDrawerItemClicked(menuItem)
    }

    override fun configureSubscriptions() {
        super.configureSubscriptions()
        if (navigation_view != null) {
            subscriptions.add(drawerAdapter
                .viewClickSubject
                .subscribe { view: View ->
                    val newPosition = navigation_recycler_view.getChildAdapterPosition(view)
                    drawerAdapter?.notifyItemChanged(drawerAdapter.selectedPosition)
                    drawerAdapter?.selectedPosition=newPosition
                    drawerAdapter?.notifyItemChanged(newPosition)
                    val itemMenu = drawerMenus[newPosition]
                    onDrawerItemClicked(itemMenu)
                }
            )

            val headerView = navigation_view.getHeaderView(0)
            headerView?.findViewById<ConstraintLayout>(R.id.cl_nav_header_home)?.setOnClickListener {
                onDrawerItemClicked(DrawerMenu(R.id.menu_profile_settings,R.string.profile_settings_label,null,ITEM))
            }

        }
    }

    open fun onDrawerItemClicked(menuItem: DrawerMenu) {
        onDrawerItemSelected(menuItem)
        closeDrawer()
    }

    open fun onDrawerItemSelected(menuItem: DrawerMenu) {}
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        if (drawerToggle != null) {
            drawerToggle?.syncState()
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (drawerToggle != null) {
            drawerToggle?.onConfigurationChanged(newConfig)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (drawerToggle?.onOptionsItemSelected(item)) {
            true
        } else super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (isDrawerOpen) {
            closeDrawer()
        } else {
            super.onBackPressed()
        }
    }

    private val isDrawerOpen: Boolean
        private get() = drawer_layout != null && navigation_view != null &&
                drawer_layout.isDrawerOpen(navigation_view)

    fun closeDrawer() {
        if (drawer_layout != null && navigation_view != null) {
            drawer_layout.closeDrawer(navigation_view)
        }
    }

    fun openDrawer() {
        if (drawer_layout != null && navigation_view != null) {
            drawer_layout.openDrawer(navigation_view)
        }
    }

    override val activityLayoutId
        protected get() = R.layout.activity_base_drawer

//    protected abstract val defaultSelectedItemId: Int
    protected fun setBaseDrawerMenuWithRide() {
        isMenuShownForRide = true
        drawerMenus = DrawerMenu.Companion.menusWithRide()
        if (navigation_recycler_view != null && drawerAdapter != null) {
            drawerAdapter?.setDrawerMenuList(drawerMenus)
            drawerAdapter?.notifyDataSetChanged()
        }
    }

    protected fun setBaseDrawerMenuWithoutRide() {
        isMenuShownForRide = false
        drawerMenus = DrawerMenu.Companion.menusWithoutRide()
        if (navigation_recycler_view != null && drawerAdapter != null) {
            drawerAdapter?.setDrawerMenuList(drawerMenus)
            drawerAdapter?.notifyDataSetChanged()
        }
    }

    protected fun setReservationsNumber(reservationCount:Int){
        drawerAdapter?.setReservationsNumber(reservationCount)
        drawerAdapter?.notifyDataSetChanged()
    }

    override fun onInternetConnectionChanged(isConnected: Boolean) {
        if (isConnected) {
            drawerToggle?.isDrawerIndicatorEnabled = true
        } else {
            drawerToggle?.isDrawerIndicatorEnabled = false
        }
        drawerToggle?.syncState()
    }
}